torchao/prototype/scaled_grouped_mm/kernels/jagged_float8_scales.py

```diff
@@ -0,0 +1,356 @@
+        - offsets: end index for each group/subtensor along dim 0
+        - output_dtype: desired float8 dtype for the output tensor
+        - round_scales_to_power_of_2: boolean indicating if scales should be rounded
+            down to the nearest power of 2.
+    Returns:
+        - float8 tensor
+        - jagged rowwise scales (i.e., rowwise scales for each group)
+    """
+    assert hp_tensor.ndim == 2, "input tensor must be 2D"
+    assert hp_tensor.is_contiguous(), "input tensor must be contiguous"
+
+    num_elements = hp_tensor.numel()
+    tl_input_dtype = FP8_DTYPE_MAP[hp_tensor.dtype]
+    tl_output_dtype = FP8_DTYPE_MAP[output_dtype]
+
+    fp8_dtype_min = torch.finfo(output_dtype).min
+    fp8_dtype_max = torch.finfo(output_dtype).max
+
+    m, k = hp_tensor.shape
+    n_groups = offsets.numel()
+
+    # allocate on-device buffers for output and scales
+    output_buffer = torch.empty_like(
+        hp_tensor, dtype=output_dtype, device=hp_tensor.device
+    )
+    scales_buffer = torch.empty(
+        (m * n_groups), dtype=torch.float32, device=hp_tensor.device
+    )
+
+    # parallelize across rows and groups (offsets)
+    grid = lambda meta: (
+        triton.cdiv(m, meta["BLOCK_SIZE_ROWS"]),
+        offsets.numel(),
+    )
+    _triton_fp8_row_major_jagged_rowwise_scales[grid](
+        hp_tensor,
+        offsets,
+        output_buffer,
+        scales_buffer,
+        m,
+        k,
+        hp_tensor.stride(0),
+        hp_tensor.stride(1),
+        output_buffer.stride(0),
+        output_buffer.stride(1),
+        num_elements,
+        fp8_dtype_min,
+        fp8_dtype_max,
+        tl_input_dtype,
+        tl_output_dtype,
+        round_scales_to_power_of_2,
+        EPS=EPS,
+    )
+    return output_buffer, scales_buffer
+
+
+@triton.autotune(configs=kernel_configs_2D, key=["num_elements"])
+@triton.jit
+def _triton_fp8_row_major_jagged_rowwise_scales(
+    input_ptr,
+    offsets_ptr,
+    out_ptr,
+    scales_ptr,
+    M: int,
+    K: int,
+    stride_input_row: int,
+    stride_input_col: int,
+    stride_output_row: int,
+    stride_output_col: int,
+    num_elements: int,
+    fp8_dtype_min: tl.constexpr,
+    fp8_dtype_max: tl.constexpr,
+    input_dtype: tl.constexpr,
+    output_dtype: tl.constexpr,
+    round_scales_to_power_of_2: tl.constexpr,
+    BLOCK_SIZE_ROWS: tl.constexpr,
+    BLOCK_SIZE_COLS: tl.constexpr,
+    EPS: tl.constexpr,
+):
+    # parallel across rows and groups (offsets)
+    block_row_id = tl.program_id(axis=0)
+    offset_idx = tl.program_id(axis=1)
+
+    # determine start and end column idx for this group
+    group_col_start_idx = tl.load(
+        offsets_ptr + offset_idx - 1, mask=offset_idx > 0, other=0
+    )
+    group_col_end_idx = tl.load(offsets_ptr + offset_idx)
+    block_row_offs = block_row_id * BLOCK_SIZE_ROWS + tl.arange(0, BLOCK_SIZE_ROWS)
+
+    # compute rowwise amaxes for this group
+    amax_buffer = tl.zeros((BLOCK_SIZE_ROWS,), dtype=tl.float64)
```

Comment:
have we kept this in fp64 elsewhere?

Reply:
hmm good point, in the float8 training code the amax is in original precision (bf16/fp32) and we convert the amax to fp64 when calculating the scale: https://github.com/pytorch/ao/blob/0045d885521aac773b89740195ca8c7d78bbc16e/torchao/float8/float8_utils.py#L45

Updated the kernel to do the same here for consistency and verified tests are still passing.